### DISCUSSION TITLE
why launch and debug with cmake tools?

### DISCUSSION THREAD

**Discussion by florenso** [microsoft/vscode-cmake-tools#3989] [Q&A] [answered]
I added some args to my launch.json. When i launch or debug my application the arguments sometimes missing...
so, there are multiple buttons in vscode to launch or debug an application. The `quick launch/debug` does not use the settings in launch.json. I tried to add some argument to the cmake tools launch button, but that also looks like it’s impossible.

why would you want to launch an application with cmake? it is just a build tool in my eyes. your launch.json could be an output of your cmake, but why would cmake tools launch of debug the application?

and why cant i disable the quick launch/debug feature?

**Comment by bobbrow** [comment, marked as answer]
The quick debug exists for simple cases where you just need to launch a build target and don't need or want to configure a whole `launch.json` file. You can customize the behavior for quick launch with the `cmake.debugConfig` setting.  That object supports any of the properties that the `launch.json` configuration would. For example:

```json
"cmake.debugConfig": {
  "args": [ "arg1", "arg2" ]
}
```

If you want to hide the buttons, use this setting:
```json
"cmake.options.advanced": {
  "launch": {
    "statusBarVisibility": "hidden"
  },
  "debug": {
    "statusBarVisibility": "hidden"
  }
}

```

**Reply by florenso** [reply]
Thanks for the quick response!

I know i can hide the status buttons but i want to fully remove the ability to launch from cmake, i do not want cmake to have the ability to run or launch my application. I would like to prevent this case #2999

**Reply by bobbrow** [reply]
Are you saying that you want to disable the commands too?  We have an API that allows for other extensions to disable the command and button if they provide similar services that are more tailored to the project's needs than our general purpose command.  But we did not make that available generally.  I don't actively work on this extension anymore, but at the time I did, this scenario was not prioritized.  (and we weren't confident that it was something we wanted to enable at the time)

**Reply by florenso** [reply]
I dont know what you mean with "commands". But i find is verry confusing that there are multiple launch/debug buttons in vscode.

To have a option that CMake tools is only used for building and not for launching, debugging, testing is a propper setting in my eyes.
IMHO i dont like that there are multiple places for launch configuration. it should be just launch.json or CMake tools, but never both.
I created an issue on this topic #3990

**Reply by bobbrow** [reply]
I'm referring to the commands in the Command Pallete.

![image](https://github.com/user-attachments/assets/1e1eeb5b-7473-4d5e-9987-839454065f03)

**Reply by florenso** [reply]
aah, yes. i would also like that the be disabled.

same for Tests, Debug and Launch in this screen.
![image](https://github.com/user-attachments/assets/fd4aa80c-513f-4f3d-a61d-d6350a0d535c)

or would you advise to ditch the whole launch.json and only configure the cmake.debugConfig in settings.json?

**Reply by bobbrow** [reply]
`launch.json` gives you the option to create multiple debug configurations so it is more powerful. If you want to have configurations that pass in different `args`, for example, you can't do that with `cmake.debugConfig` unless you constantly update the setting. But the reason the quick debug feature exists is because most people can't be bothered to write a `launch.json` file so I don't expect that it's going away any time soon.

For disabling the other places where quick debug appears, please make sure you include that in your feature request.

**Reply by florenso** [reply]
That explains it a bit more, i will add that in the feature request.

I also i noticed there is no cmake.launchConfig. is there a way i can configure a normal launch with cmake tools?

**Reply by bobbrow** [reply]
launch will use the same info from `cmake.debugConfig`. The schema is the same for debug and launch, so a separate setting was not added.
